design/legacy-block-storage.md

```diff
@@ -0,0 +1,141 @@
+* Many expensive and proprietary appliance storage silos are still providing services for Enterprise IT infrastructures.
+* Many companies only trust well-certified and time-proven storage solutions for their mission critical applications
+
+For transitioning into existing enterprise IT infrastructures, Rook will have a much easier time with support for legacy block storage systems.
+
+
+### Storage Resource Normalization
+
+The OpenStack Block Storage service (Cinder) provides a standardized architecture to aggregate the features and management of all kinds of block storage systems.
+It is the first service to have a standardized interface that is widely supported by most storage vendors including Dell EMC, IBM, NetApp, Hitachi, HPE, Pure Storage, VMware, etc.
+
+The major functions of the Cinder volume driver include:
+
+* Create/Delete Volume
+* Attach/Detach Volume
+* Extend Volume
+* Create/Delete Snapshot
+* Create Volume from Snapshot
+* Create Volume from Volume (clone)
+* Create Image from Volume
+* Volume Migration (host assisted)
+
+The following are some references for OpenStack Cinder:
+
+* [OpenStack Cinder](https://access.redhat.com/documentation/en/red-hat-enterprise-linux-openstack-platform/version-7/red-hat-enterprise-linux-openstack-platform-7-architecture-guide/chapter-1-components#comp-cinder)
+* [Cinder Driver Support Matrix](https://wiki.openstack.org/wiki/CinderSupportMatrix)
+
+In this project, we propose to leverage the open source project [OpenStack Cinder](https://github.com/openstack/cinder) as the unified storage connector that integrates legacy block storage with Rook.
+Through this implementation, we can quickly extend Rook's storage backend support to most modern block storage solutions that are already used in production today.
+
+
+## Architecture
+
+![legacy-block-storage-architecture-diagram](legacy-block-storage-architecture.png)
+
+### Operator
+
+* start and monitor the Cinder service pods
+* watch for state changes and apply the changes
+* monitor the health of external storage systems
+* manage custom resource definitions (CRDs) for pools
+
+
+### Agent
+
+* deployed on each Kubernetes nodes
+* handle attaching/detaching, mounting/unmounting volumes and formatting the file system
+
+
+## Details
+
+### lbs (legacy block storage) cluster CRD example
+
+```yaml
+apiVersion: lbs.rook.io/v1alpha1
+kind: Cluster
+metadata:
+  name: rook-lbs
+  namespace: rook-lbs
+spec:
+  cinderVersion:
+    image: openstack/cinder:v13.0.2-4
+  dataDirHostPath: /var/lib/rook
+  api:
+    count: 3
+    allowMultiplePerNode: true
+  storage:
+    useAllNodes: true
+    resources:
+      limits:
+        cpu: 2
+        memory: 2048Mi
+      requests:
+        cpu: 2
+        memory: 2048Mi
+    config:
+      enabled_backends: svc1234, netapp123
+      svc1234:
+        volume_backend_name: svc1234
+        volume_driver: cinder.volume.drivers.ibm.storwize_svc.storwize_svc_iscsi.StorwizeSVCISCSIDriver
+        san_ip: 1.2.3.4
```

Comment:
I see the possibility for some complication from looking at the CRD, though this may not be a true issue.

I see `san_ip` here and a corresponding `netapp_server_hostname` below. Writing handling code in the lbs operator for custom fields for all the different volume drivers would be a lot of work, and I'm not sure how these fields might change for different versions of Cinder. It could become a lot of busy work to support in the long term if the operator needs to understand all of the different fields.

Are these values passed directly to the cinder pod by the operator without special handling needed? If that is the case, my concerns are much assuaged. In this case, my thoughts revolve mostly around making sure the user sees a useful error message from the lbs operator like `user inputs appear to be invalid. [last error message from cinder]. see pod ABCDEFG-WXYZ logs for detailed logs.` to help them debug the config error.

Please clarify this implementation detail in the proposal, and I would suggest that the error message also be included in the proposal, as designing how the user can debug failures is an important and all-too-often overlooked part of software design.

Tangential thought: If these fields directly correspond to commandline flags, keep in mind that in yaml, hyphens are valid in key names. `--san-ip` may be a commandline flag, for example, and `san-ip` is a valid yaml key.